test/test_oversized_bag.cpp

```diff
@@ -0,0 +1,222 @@
+    static_assert(std::is_same_v<decltype(bbag)::size_type, size_t>);
+    static_assert(std::is_same_v<decltype(bbag)::for_all_args,
+                                 std::tuple<decltype(bbag)::value_type>>);
+  }
+
+  //
+  // Test Rank 0 async_insert
+  {
+    ygm::container::oversized_bag<std::string> bbag(world);
+    if (world.rank0()) {
+      bbag.async_insert("dog");
+      bbag.async_insert("apple");
+      bbag.async_insert("red");
+    }
+    world.barrier();
+
+    YGM_ASSERT_RELEASE(bbag.count("dog") == 1);
+    YGM_ASSERT_RELEASE(bbag.count("apple") == 1);
+    YGM_ASSERT_RELEASE(bbag.count("red") == 1);
+    YGM_ASSERT_RELEASE(bbag.size() == 3);
+    bbag.clear();
+  }
+
+  //
+  // Test Rank 0 local insert on addition past file threshold
+  {
+    size_t threshold = 1;
+    ygm::container::oversized_bag<std::string> bbag = ygm::container::oversized_bag<std::string>(world, threshold);
+    if (world.rank0()) {
+      bbag.local_insert("dog");
+      YGM_ASSERT_RELEASE(bbag.num_local_files() == 1);
+      bbag.local_insert("apple");
+      YGM_ASSERT_RELEASE(bbag.num_local_files() == 2);
+      bbag.local_insert("red");
+      YGM_ASSERT_RELEASE(bbag.num_local_files() == 3);
+    }
+    world.barrier();
+    
+    // Count and size functions come from overarching ygm container,
+    // rely on local functions
+    YGM_ASSERT_RELEASE(bbag.count("dog") == 1);
+    YGM_ASSERT_RELEASE(bbag.count("apple") == 1);
+    YGM_ASSERT_RELEASE(bbag.count("red") == 1);
+    YGM_ASSERT_RELEASE(bbag.size() == 3);
+    bbag.clear();
+  }
+
+  //
+  // Test all ranks async_insert
+  {
+    ygm::container::oversized_bag<std::string> bbag(world);
+    bbag.async_insert("dog");
+    bbag.async_insert("apple");
+    bbag.async_insert("red");
+    YGM_ASSERT_RELEASE(bbag.size() == 3 * (size_t)world.size());
+    YGM_ASSERT_RELEASE(bbag.count("dog") == (size_t)world.size());
+    YGM_ASSERT_RELEASE(bbag.count("apple") == (size_t)world.size());
+    YGM_ASSERT_RELEASE(bbag.count("red") == (size_t)world.size());
+
+    {
+      std::vector<std::string> all_data;
+      bbag.gather(all_data, 0);
+      if (world.rank0()) {
+        YGM_ASSERT_RELEASE(all_data.size() == 3 * (size_t)world.size());
+      }
+    }
+    {
+      std::set<std::string> all_data;
+      bbag.gather(all_data, 0);
+      if (world.rank0()) {
+        YGM_ASSERT_RELEASE(all_data.size() == 3);
+      }
+    }
+  }
+
+  //
+  // Test Rebalance
+  {
+    ygm::container::oversized_bag<std::string> bbag(world);
+    bbag.async_insert("begin", 0);
+    bbag.async_insert("end", world.size() - 1);
+    bbag.rebalance();
+    YGM_ASSERT_RELEASE(bbag.local_size() == 2);
+  }
+
+  // Currently, shuffle functions are being ignored as per Roger
+  // Test local_shuffle and global_shuffle
+  // {
+  //   ygm::container::oversized_bag<int> bbag(world);
+  // 
+  // }
+
+  // 
+  // Test for_all
+  {
+    ygm::container::oversized_bag<std::string> bbag(world);
```

Comment:
Same as above, the previous state will be persisted even across creating a 'new' local oversized bag